File: src/parser.rs
use logos::{Lexer, Logos, SpannedIter};
use std::fmt;
use std::num::ParseIntError;

// NOTE:
// See https://lalrpop.github.io/lalrpop/lexer_tutorial/005_external_lib.html
// for a tutorial.
// TODO:
// when rust support derive fields to be constant we may use this:
// use pomsky_macro::pomsky;

#[derive(Default, Debug, Clone, PartialEq)]
pub enum LexicalError {
    InvalidInteger(ParseIntError),
    #[default]
    InvalidToken,
}

impl From<ParseIntError> for LexicalError {
    fn from(err: ParseIntError) -> Self {
        LexicalError::InvalidInteger(err)
    }
}

#[derive(Logos, Clone, Debug, PartialEq, Hash, Eq)]
// BUG: comments are not nestable for now
#[logos(skip r"[ \t\f]+|#(\n|[^\|][^\n]*)|#\|([^|]|\|[^#])*\|?\|#", error = LexicalError)]
pub enum Token<'input> {
    #[regex(r"\r|\n|\r\n")]
    NL,

    #[regex("-?(0|[1-9][0-9]*)")]
    DecimalLit(&'input str),

    #[regex("0x[0-9a-fA-F]+")]
    HexLit(&'input str),

    // #[regex(r"#(\n|[^\|][^\n]*)")]

    // #[regex("[_a-zA-Z][_0-9a-zA-Z]*", |lex| lex.slice().to_string())]
    // Identifier(&'input str),
    #[regex(r"'([^'\\]|\\['\\])*'")]
    LiteralString(&'input str),

    // TODO: support templated strings in multiline string
    #[regex(r"\\\\\\[^\n]*")]
    MultilineStringPart(&'input str),

    // TODO: support templated string to contain arbitrary expression
    #[regex(r#""([^"\\{}]|\\["\\{}$]|\$[a-z]*)*""#)]
    TemplStr(&'input str),
    // NOTE: the below won't work as it creates conflicts. We may actually need to invoke the
    // parser multiple times.

    // #[regex(r#""([^"\\{}]|\\["\\{}$]|\$[a-z]*)*\$\{"#)]
    // TemplStrHead,
    // #[regex(r#"}([^"\\{}]|\\["\\{}$]|\$[a-z]*)*""#)]
    // TemplStrEnd,
    // #[regex(r#"}([^"\\{}]|\\["\\{}$]|\$[a-z]*)*\$\{"#)]
    // TemplStrMid,
    #[regex(r"~[a-z][a-zA-Z0-9_]*")]
    SigilIdent(&'input str),

    #[regex(r"~[a-z]\(([^)\\]|\\[)\\])*\)")]
    SigilParened(&'input str),

    #[regex(r"~[a-z]\[([^]\\]|\\[\]\\])*\]")]
    SigilBracketed(&'input str),

    #[regex(r"~[a-z]\{([^}\\]|\\[}\\])*\}")]
    SigilBraced(&'input str),

    #[regex(r"/([^/\\]|\\[\\/])*/[a-z]*")]
    RegexLike(&'input str),

    // `=` is forbidden because we want assignments
    #[regex(r"[^= \r\n\t\f()]+", priority = 1)]
    Word(&'input str),

    #[token("){")]
    RPARENLBRACE,
    #[token("()")]
    UNIT,
    #[token("(")]
    LPAREN,
    #[token(")")]
    RPAREN,
    #[token("[")]
    LBRACKET,
    #[token("]")]
    RBRACKET,
    #[token("{")]
    LBRACE,
    #[token("}")]
    RBRACE,

    #[token("if")]
    IF,
    #[token("else")]
    ELSE,
    #[token("for")]
    FOR,
    #[token("in")]
    IN,
    #[token("&&")]
    ANDTHEN,
    #[token("||")]
    ORELSE,
    #[token("=")]
    ASSIGN,
    #[token("|")]
    PIPE,
    #[token(",")]
    COMMA,
    #[token(":")]
    COLON,
    #[token(";")]
    SEMICOLON,
    #[token("+")]
    PLUS,
    #[token("-")]
    MINUS,
    #[token("*")]
    MULT,
    #[token("/")]
    DIV,
    #[token("%")]
    REM,
    #[token("==")]
    EQ,
    #[token("!=")]
    NEQ,
    #[token("<=")]
    LEQ,
    #[token(">=")]
    GEQ,
    #[token("<")]
    LT,
    #[token(">")]
    GT,
    #[token("and")]
    AND,
    #[token("or")]
    OR,
    #[token("<<")]
    BSHIFTL,
    #[token(">>")]
    BSHIFTR,
    #[token("^")]
    BXOR,
    #[token("&")]
    BAND,
    #[token("++")]
    CONCAT,
}

impl<'input> fmt::Display for Token<'input> {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(f, "{:?}", self)
    }
}

pub type Spanned<Tok, Loc, Error> = Result<(Loc, Tok, Loc), Error>;

pub struct WrappedLexer<'input> {
    // instead of an iterator over characters, we have a token iterator
    token_stream: SpannedIter<'input, Token<'input>>,
}

impl<'input> WrappedLexer<'input> {
    pub fn new(input: &'input str) -> Self {
        // the Token::lexer() method is provided by the Logos trait
        Self {
            token_stream: Token::lexer(input).spanned(),
        }
    }
}

impl<'input> Iterator for WrappedLexer<'input> {
    type Item = Spanned<Token<'input>, usize, LexicalError>;

    fn next(&mut self) -> Option<Self::Item> {
        self.token_stream
            .next()
            .map(|(token, span)| Ok((span.start, token?, span.end)))
    }
}

#[cfg(test)]
mod tests {
    use crate::{ast::Expr, grammar::ProgramParser, parser::WrappedLexer};

    fn try_parse(input: &str) -> Option<Vec<Expr>> {
        let lexer = WrappedLexer::new(input);
        let parser = ProgramParser::new();
        Some(parser.parse(lexer).unwrap())
    }
    fn assert_parse(input: &str, parsed: &str) {
        assert_eq!(format!("{:?}", try_parse(input).unwrap()), parsed);
    }

    fn try_parse_verbose(input: &str) -> Option<Vec<Expr>> {
        let lexer = WrappedLexer::new(input);
        let parser = ProgramParser::new();
        let lexer2 = WrappedLexer::new(input);
        for tok in lexer2 {
            println!("{:?}", tok);
        }
        Some(parser.parse(lexer).unwrap())
    }
    fn assert_parse_verbose(input: &str, parsed: &str) {
        assert_eq!(format!("{:?}", try_parse_verbose(input).unwrap()), parsed);
    }

    #[test]
    fn parse_single_line_comment() {
        assert_parse(
            "echo 1 #hello I'm here
            echo 2 #wow\n",
            "[Command(\"echo\", [Val(VStrStatic(\"1\"))], []), Command(\"echo\", [Val(VStrStatic(\"2\"))], [])]",
        );
    }

    // TODO: make it nestable, need to rewrite the lexer.
    #[test]
    fn parse_multi_line_comment() {
        assert_parse(
            " #| hello I'm here with multiline comments |#

            echo 2 #wow\n",
            "[Command(\"echo\", [Val(VStrStatic(\"2\"))], [])]",
        );
    }

    #[test]
    fn parse_exp_simple() {
        assert_parse("(1)\n", "[Val(VInt(1))]");
        assert_parse("(((((((())))))))\n", "[Val(VUnit)]");
        assert_parse(
            "('This ACTUALLY works!')\n",
            "[Val(VStr(\"This ACTUALLY works!\"))]",
        );

        assert_parse("(1 + 1)\n", "[Binary(Val(VInt(1)), Plus, Val(VInt(1)))]");
    }

    #[test]
    fn parse_exp_complex() {
        assert_parse(
            "(1 + 1 * 3 << 3 | 9 ^ 100 and '1' ++ '2' == '12')\n",
            "[Binary(Binary(Binary(Binary(Val(VInt(1)), Plus, Binary(Val(VInt(1)), Mult, Val(VInt(3)))), Bshiftl, Val(VInt(3))), Pipe, Binary(Val(VInt(9)), Bxor, Val(VInt(100)))), And, Binary(Binary(Val(VStr(\"1\")), Concat, Val(VStr(\"2\"))), Eq, Val(VStr(\"12\"))))]",
        );
    }

    #[test]
    fn parse_cmds() {
        assert_parse("ls\n\n", r#"[Command("ls", [], [])]"#);
        assert_parse(
            "cd ..\npwd\n",
            "[Command(\"cd\", [Val(VStrStatic(\"..\"))], []), Command(\"pwd\", [], [])]",
        );
        assert_parse(
            "test -d my.zip && 7z x my.zip || echo no zip file\n",
            "[Binary(Binary(Command(\"test\", [Val(VStrStatic(\"-d\")), Val(VStrStatic(\"my.zip\"))], []), AndThen, Command(\"7z\", [Val(VStrStatic(\"x\")), Val(VStrStatic(\"my.zip\"))], [])), OrElse, Command(\"echo\", [Val(VStrStatic(\"no\")), Val(VStrStatic(\"zip\")), Val(VStrStatic(\"file\"))], []))]",
        );
        assert_parse(
            "ls 1 2 3 'some string' -999 -la\n",
            "[Command(\"ls\", [Val(VStrStatic(\"1\")), Val(VStrStatic(\"2\")), Val(VStrStatic(\"3\")), Val(VStr(\"some string\")), Val(VStrStatic(\"-999\")), Val(VStrStatic(\"-la\"))], [])]",
        );
        assert_parse(
            "ls && 7z && mkfs.ntfs && x86_64-pc-linux-gnu-c++-11\n",
            "[Binary(Binary(Binary(Command(\"ls\", [], []), AndThen, Command(\"7z\", [], [])), AndThen, Command(\"mkfs.ntfs\", [], [])), AndThen, Command(\"x86_64-pc-linux-gnu-c++-11\", [], []))]",
        );
    }

    #[test]
    fn parse_if() {
        assert_parse(
            "if (1 + 1 == 2) { echo yes } else { echo world ends. }",
            "[If(Binary(Binary(Val(VInt(1)), Plus, Val(VInt(1))), Eq, Val(VInt(2))), Block([Command(\"echo\", [Val(VStrStatic(\"yes\"))], [])]), Block([Command(\"echo\", [Val(VStrStatic(\"world\")), Val(VStrStatic(\"ends.\"))], [])]))]",
        );
    }

    #[test]
    fn parse_assign() {
        assert_parse(
            "a = 1 + 1\n",
            "[Binary(Ident(\"a\"), Assign, Binary(Val(VInt(1)), Plus, Val(VInt(1))))]",
        );
    }
}

File: lib/parser.ml
include Nice_parser.Make(struct
  type result = Ast.expr
  type token = Menhir_parser.token
  exception ParseError = Menhir_parser.Error
  let parse = Menhir_parser.program_eof
  include Lexer
end)

(*===========================================================================*)
(* TESTS                                                                     *)
(* https://dune.readthedocs.io/en/stable/tests.html#inline-expectation-tests *)
(* https://github.com/smolkaj/nice-parser/blob/master/example/parser.ml      *)
(*===========================================================================*)

let%test_module _ = (module struct
  
  let () = Printexc.record_backtrace false

  let%expect_test "single-line-comment" =

    parse_string {|
      echo 1 #hello I'm here
      echo 2 #wow
    |}
    |> Printf.printf !"%{sexp:Ast.expr}";
    [%expect{| (MBlock ((Command echo ((Just 1)) ()) (Command echo ((Just 2)) ()))) |}]

  let%expect_test "nestable-comments" =
    parse_string {|
      #| hello I'm here with #|nestable|# comments |#
      echo 1
    |}
    |> Printf.printf !"%{sexp:Ast.expr}";
    [%expect{| (MBlock ((Command echo ((Just 1)) ()))) |}]

  let%expect_test "cmd-simple" =
    parse_string "pwd"
    |> Printf.printf !"%{sexp:Ast.expr}";
    [%expect{| (MBlock ((Command pwd () ()))) |}]

  let%expect_test "exp-int" =
    parse_string "= 1"
    |> Printf.printf !"%{sexp:Ast.expr}";
    [%expect{| (MBlock ((Val (Int 1)))) |}]

  let%expect_test "exp-unit-nested" =
    parse_string "= ((((((()))))))"
    |> Printf.printf !"%{sexp:Ast.expr}";
    [%expect{| (MBlock ((Val Unit))) |}]

  let%expect_test "exp-string" =
    parse_string "= 'This ACTUALLY works!'"
    |> Printf.printf !"%{sexp:Ast.expr}";
    [%expect{|  (MBlock ((Val (Str "This ACTUALLY works!"))))  |}]

  let%expect_test "exp-simple" =
    parse_string "= 1 + 1"
    |> Printf.printf !"%{sexp:Ast.expr}";
    [%expect{|
      (MBlock ((Binary Add (Val (Int 1)) (Val (Int 1)))))
    |}]

  let%expect_test "exp-complex" =
    parse_string "(1 + 1 * 3 shl 3 bor 9 bxor 100 and '1' ++ '2' == '12')"
    |> Printf.printf !"%{sexp:Ast.expr}";
    [%expect{|
      (MBlock
       ((Binary LAnd
         (Binary BOr
          (Binary BShiftL
           (Binary Add (Val (Int 1)) (Binary Mul (Val (Int 1)) (Val (Int 3))))
           (Val (Int 3)))
          (Binary BXor (Val (Int 9)) (Val (Int 100))))
         (Binary Eq (Binary SConcat (Val (Str 1)) (Val (Str 2))) (Val (Str 12))))))
    |}]

  let%expect_test "simple cmd" =
    parse_string "test -d my.zip && 7z x my.zip || echo no zip file"
    |> Printf.printf !"%{sexp:Ast.expr}";
    [%expect{|
        (MBlock
         ((Binary OrElse
           (Binary AndThen (Command test ((Just -d) (Just my.zip)) ()) 
            (Command 7z ((Just x) (Just my.zip)) ()))
           (Command echo ((Just no) (Just zip) (Just file)) ()))))
    |}]

  let%expect_test "different word types" =
    parse_string "ls 1 2 3 'some string' -999 -la"
    |> Printf.printf !"%{sexp:Ast.expr}";
    [%expect{|
      (MBlock
       ((Command ls
         ((Just 1) (Just 2) (Just 3) (Just "some string") (Just -999) (Just -la))
         ())))
    |}]


  let%expect_test "weird cmds" =
    parse_string "ls && 7z && mkfs.ntfs && x86_64-pc-linux-gnu-c++-11"
    |> Printf.printf !"%{sexp:Ast.expr}";
    [%expect{|
        (MBlock
         ((Binary AndThen
           (Binary AndThen (Binary AndThen (Command ls () ()) (Command 7z () ()))
            (Command mkfs.ntfs () ()))
           (Command x86_64-pc-linux-gnu-c++-11 () ()))))
    |}]

  let%expect_test "if" =
    parse_string "if (1 + 1 == 2) { echo yes } else { echo world ends. }"
    |> Printf.printf !"%{sexp:Ast.expr}";
    [%expect{|
        (MBlock
         ((If (Binary Eq (Binary Add (Val (Int 1)) (Val (Int 1))) (Val (Int 2)))
           (Block ((Command echo ((Just yes)) ())))
           (Block ((Command echo ((Just world) (Just ends.)) ()))))))
    |}]

  let%expect_test "assign" =
    parse_string "a = 1 + 1"
    |> Printf.printf !"%{sexp:Ast.expr}";
    [%expect{|
      (MBlock ((Assign a (Binary Add (Val (Int 1)) (Val (Int 1))) false)))
    |}]

  let%expect_test "expr as word arg" =
    parse_string "echo [1 + 1]"
    |> Printf.printf !"%{sexp:Ast.expr}";
    [%expect{|
      (MBlock ((Command echo ((Exp (Binary Add (Val (Int 1)) (Val (Int 1))))) ())))
    |}]

  let%expect_test "semicol as line delimiter" =
    parse_string "ls; pwd; echo 1"
    |> Printf.printf !"%{sexp:Ast.expr}";
    [%expect{|
      (MBlock
       ((Command ls () ()) (Command pwd () ()) (Command echo ((Just 1)) ())))
    |}]

  let%expect_test "templated string" =
    parse_string {|hey (""{1 + 1}"")|}
    |> Printf.printf !"%{sexp:Ast.expr}";
    [%expect{|
      (MBlock
       ((Command hey ()
         ((Binary SConcat (Val (Str ""))
           (Binary SConcat
            (Command stringify () ((Binary Add (Val (Int 1)) (Val (Int 1)))))
            (Val (Str ""))))))))
    |}]

  let%expect_test "multi string" =
    parse_string {|
      hey (
        \\\ hello
        \\\{ 1 + 1 }
        \\\ world
      )
    |}
    |> Printf.printf !"%{sexp:Ast.expr}";
    [%expect{|
      (MBlock
       ((Command hey ()
         ((Binary SConcat
           (Binary SConcat (Val (Str " hello"))
            (Binary Add (Val (Int 1)) (Val (Int 1))))
           (Val (Str " world")))))))
    |}]

end)
